sql/planner/tree.go

```diff
@@ -0,0 +1,509 @@
+
+	toStream(st document.Stream) (document.Stream, error)
+}
+
+type outputNode interface {
+	Node
+
+	toResult(st document.Stream) (query.Result, error)
+}
+
+type node struct {
+	op          Operation
+	left, right Node
+}
+
+func (n *node) Operation() Operation {
+	return n.op
+}
+
+func (n *node) Left() Node {
+	return n.left
+}
+
+func (n *node) Right() Node {
+	return n.right
+}
+
+func (n *node) SetLeft(ln Node) {
+	n.left = ln
+}
+
+func (n *node) SetRight(rn Node) {
+	n.right = rn
+}
+
+func (n *node) Equal(other Node) bool {
+	if n == nil {
+		return other == nil
+	}
+
+	if other == nil {
+		return false
+	}
+
+	if n.op != other.Operation() {
+		return false
+	}
+
+	if n.left == nil {
+		if other.Left() != nil {
+			return false
+		}
+	} else if !n.left.Equal(other.Left()) {
+		return false
+	}
+
+	if n.right == nil {
+		if other.Right() != nil {
+			return false
+		}
+	} else if !n.right.Equal(other.Right()) {
+		return false
+	}
+
+	return true
+}
+
+type selectionNode struct {
+	node
+
+	cond   expr.Expr
+	tx     *database.Transaction
+	params []expr.Param
+}
+
+var _ operationNode = (*selectionNode)(nil)
+
+// NewSelectionNode creates a node that filters documents of a stream, according to
+// the expression condition.
+func NewSelectionNode(n Node, cond expr.Expr) Node {
+	return &selectionNode{
+		node: node{
+			op:   Selection,
+			left: n,
+		},
+		cond: cond,
+	}
+}
+
+func (n *selectionNode) Equal(other Node) bool {
+	if !n.node.Equal(other) {
+		return false
+	}
+
+	return n.cond.Equal(other.(*selectionNode).cond)
+}
+
+func (n *selectionNode) Bind(tx *database.Transaction, params []expr.Param) (err error) {
+	n.tx = tx
+	n.params = params
+	return
+}
+
+func (n *selectionNode) toStream(st document.Stream) (document.Stream, error) {
+	if n.cond == nil {
+		return st, nil
+	}
+
+	stack := expr.EvalStack{
+		Tx:     n.tx,
+		Params: n.params,
+	}
+
+	return st.Filter(func(d document.Document) (bool, error) {
+		stack.Document = d
+		v, err := n.cond.Eval(stack)
+		if err != nil {
+			return false, err
+		}
+
+		return v.IsTruthy(), nil
+	}), nil
+}
+
+type limitNode struct {
+	node
+
+	limitExpr expr.Expr
+	tx        *database.Transaction
+	params    []expr.Param
+}
+
+var _ operationNode = (*limitNode)(nil)
+
+// NewLimitNode creates a node that limits the number of documents processed by the stream.
+func NewLimitNode(n Node, limitExpr expr.Expr) Node {
+	return &limitNode{
+		node: node{
+			op:   Limit,
+			left: n,
+		},
+		limitExpr: limitExpr,
+	}
+}
+
+func (n *limitNode) Equal(other Node) bool {
+	if !n.node.Equal(other) {
+		return false
+	}
+
+	return n.limitExpr.Equal(other.(*limitNode).limitExpr)
+}
+
+func (n *limitNode) Bind(tx *database.Transaction, params []expr.Param) (err error) {
+	n.tx = tx
+	n.params = params
+	return
+}
+
+func (n *limitNode) toStream(st document.Stream) (document.Stream, error) {
+	stack := expr.EvalStack{
+		Tx:     n.tx,
+		Params: n.params,
+	}
+
+	v, err := n.limitExpr.Eval(stack)
```

Comment:
We should check that the `limitExpr` is not nil before evaluating it.

Reply:
It's already checked before hand, the node cannot be created if the limitExpr is nil. I'm removing the limitExpr in another branch anyway 👍

Reply:
I have cherry-picked the commit and added it to this PR